### ISSUE TITLE
ES Executor Auto-Discovery Not Working with Multi-Node VirtualBox Configuration

### ISSUE THREAD

**Issue by dvonthenen** [mesos/elasticsearch#517]
Hi,

I have been trying to get the ES Framework working on a multi-node VirtualBox configuration that is being driven by [Playa-Mesos](https://github.com/emccode/vagrant/tree/master/playa-mesos). Playa-Mesos uses vagrant to deploy the infrastructure and do an initial configuration of various things like Mesos, REX-Ray, and etc. The deployment of the actual infrastructure works but some post configuration needs to be run ([script here](https://raw.githubusercontent.com/dvonthenen/junkyard/master/mesos/elasticsearch/updateIP.sh)) in order for the master node to contact the scheduler on the slave node and some other small things. The end result of the Playa-Mesos deployment is a single master with 4 slave nodes and each master/slave node is on their own VirtualBox instance (5 instances total).

I invoke the ES Framework using the following command:
```
curl -k -XPOST -d @elasticsearch_ext_virtualbox.json -H "Content-Type: application/json" 10.141.141.10:8080/v2/apps
```

and the following JSON (dvonthenen/elasticsearch-scheduler-test is basically the same Docker Image as I use to demo the external storage PR functionality on AWS [here](https://github.com/mesos/elasticsearch/pull/489) except with TRACE enabled):
```
{
  "id": "elasticsearch-mesos-scheduler",
  "container": {
    "docker": {
      "image": "dvonthenen/elasticsearch-scheduler-test",
      "network": "HOST"
    }
  },
  "args": ["--zookeeperMesosUrl", "zk://10.141.141.10:2181/mesos", "--executorImage", "dvonthenen/elasticsearch-executor-test", "--executorTimeout", "3600000", "--executorHealthDelay", "3600000", "--useIpAddress", "true", "--externalVolumeDriver", "rexray"],
  "cpus": 0.5,
  "mem": 512.0,
  "env": {
    "JAVA_OPTS": "-Xms128m -Xmx256m"
  },
  "instances": 1
}
```

The Scheduler is deployed no problem, the first ES executor is deployed no problem, but the second and third nodes fail because the ES cannot auto-discover the other nodes which causes the timeout to hit and the 2nd and 3rd nodes get redeployed continuously. I am pretty much stuck at this point since I am not an Elastic Search expert. How do I go about getting this configuration to work in the auto-discover mode?

**Comment by philwinder** [comment]
Hi again David.
Aside from you're custom image (you're PR will be merged today!), your settings look fine. But you haven't submitted any ram requirements so there is a possibility that if you insert data you might get an OOM error. But I don't think you are so that is not a problem.

There is no "auto" discovery mechanism. Multicast is disabled in ES because it is disabled in AWS and GCE.

Hence, we use unicast and set the known nodes to the previously started nodes. Hence, the logic is:
- start one node
- wait for node to be green
- start rest of nodes with ip address of first node

ES will then use the gossip protocol to update the rest of the cluster once it has attached to the first node.

So, you're issue is likely due to the fact that your other Vagrant VMs are not routable. E.g. They may be firewalled, on different subnets, etc.

So basically, I am saying that this isn't a problem with elasticsearch, it's an issue with your network setup. Sorry :-)

I'll close for now, but I am happy to reopen if there is an issue.
